cpp/src/arrow/python/numpy_convert.cc

```diff
@@ -274,5 +275,348 @@ Status TensorToNdarray(const std::shared_ptr<Tensor>& tensor, PyObject* base,
+  }
+
+  const int ndim_indices = sparse_index_indices->ndim();
+  std::vector<npy_intp> npy_shape_indices(ndim_indices);
+
+  for (int i = 0; i < ndim_indices; ++i) {
+    npy_shape_indices[i] = sparse_index_indices->shape()[i];
+  }
+
+  const void* immutable_data = nullptr;
+  if (sparse_tensor->data()) {
+    immutable_data = sparse_tensor->data()->data();
+  }
+
+  const void* immutable_indptr = nullptr;
+  if (sparse_index_indptr->data()) {
+    immutable_indptr = sparse_index_indptr->data()->data();
+  }
+
+  const void* immutable_indices = nullptr;
+  if (sparse_index_indices->data()) {
+    immutable_indices = sparse_index_indices->data()->data();
+  }
+
+  // Remove const =(
+  void* mutable_data = const_cast<void*>(immutable_data);
+  void* mutable_indptr = const_cast<void*>(immutable_indptr);
+  void* mutable_indices = const_cast<void*>(immutable_indices);
+
+  int array_flags = 0;
+  if (sparse_tensor->is_mutable()) {
+    array_flags |= NPY_ARRAY_WRITEABLE;
+  }
+
+  int type_num_indptr;
+  RETURN_NOT_OK(GetNumPyType(*sparse_index_indptr->type(), &type_num_indptr));
+  PyArray_Descr* dtype_indptr = PyArray_DescrNewFromType(type_num_indptr);
+
+  int type_num_indices;
+  RETURN_NOT_OK(GetNumPyType(*sparse_index_indptr->type(), &type_num_indices));
+  PyArray_Descr* dtype_indices = PyArray_DescrNewFromType(type_num_indices);
+
+  PyObject* result_data =
+      PyArray_NewFromDescr(&PyArray_Type, dtype_data, 1, npy_shape_data.data(), nullptr,
+                           mutable_data, array_flags, nullptr);
+  PyObject* result_indptr =
+      PyArray_NewFromDescr(&PyArray_Type, dtype_indptr, 1, npy_shape_indptr.data(),
+                           nullptr, mutable_indptr, array_flags, nullptr);
+  PyObject* result_indices =
+      PyArray_NewFromDescr(&PyArray_Type, dtype_indices, 1, npy_shape_indices.data(),
+                           nullptr, mutable_indices, array_flags, nullptr);
+  RETURN_IF_PYERROR()
+
+  if (base == Py_None || base == nullptr) {
+    base = py::wrap_sparse_tensor_csr(sparse_tensor);
+  } else {
+    Py_XINCREF(base);
+  }
+  PyArray_SetBaseObject(reinterpret_cast<PyArrayObject*>(result_data), base);
+  PyArray_SetBaseObject(reinterpret_cast<PyArrayObject*>(result_indptr), base);
+  PyArray_SetBaseObject(reinterpret_cast<PyArrayObject*>(result_indices), base);
+
+  *out_data = result_data;
+  *out_indptr = result_indptr;
+  *out_indices = result_indices;
+  return Status::OK();
+}
+
+Status NdarrayToSparseTensorCOO(MemoryPool* pool, PyObject* ao,
+                                std::shared_ptr<SparseTensorCOO>* out) {
+  if (!PyArray_Check(ao)) {
+    return Status::TypeError("Did not pass ndarray object");
+  }
+
+  PyArrayObject* ndarray = reinterpret_cast<PyArrayObject*>(ao);
+
+  // TODO(wesm): What do we want to do with non-contiguous memory and negative strides?
+
+  int ndim = PyArray_NDIM(ndarray);
+
+  // This is also holding the GIL, so don't already draw it.
```

Comment:
I don't understand what this means. Apparently you're calling this function with `nogil`.